includes/REST_API/Embed_Controller.php

```
@@ -0,0 +1,209 @@
+
+		$title  = '';
+		$poster = '';
+
+		$data = [
+			'title'  => $title,
+			'poster' => $poster,
+		];
+
+		$args = [
+			'limit_response_size' => 153600, // 150 KB.
+			'timeout'             => 7, // phpcs:ignore WordPressVIPMinimum.Performance.RemoteRequestTimeout.timeout_timeout
+		];
+
+		/**
+		 * Filters the HTTP request args for link data retrieval.
+		 *
+		 * Can be used to adjust timeout and response size limit.
+		 *
+		 * @param array $args Arguments used for the HTTP request
+		 * @param string $url The attempted URL.
+		 */
+		$args = apply_filters( 'web_stories_embed_data_request_args', $args, $url );
+
+		$response = wp_safe_remote_get( $url, $args );
+
+		if ( WP_Http::OK !== wp_remote_retrieve_response_code( $response ) ) {
+			// Not saving the error response to cache since the error might be temporary.
+			return new WP_Error( 'rest_invalid_url', get_status_header_desc( 404 ), [ 'status' => 404 ] );
+		}
+
+		$html = wp_remote_retrieve_body( $response );
+
+		if ( ! $html ) {
+			set_transient( $cache_key, wp_json_encode( $data ), $cache_ttl );
+			return new WP_Error( 'rest_invalid_url', get_status_header_desc( 404 ), [ 'status' => 404 ] );
+		}
+
+		$doc                      = new DOMDocument();
+		$doc->strictErrorChecking = false; // phpcs:ignore WordPress.NamingConventions.ValidVariableName.UsedPropertyNotSnakeCase
+
+		// Suppress warnings generated by loadHTML.
+		$errors = libxml_use_internal_errors( true );
+		$doc->loadHTML( $html );
+		libxml_clear_errors();
+		libxml_use_internal_errors( $errors );
+		$xpath = new DOMXpath( $doc );
+
+		$amp_story = $xpath->query( '//amp-story' );
+		$title     = $this->get_dom_attribute_content( $amp_story, 'title' );
+		$poster    = $this->get_dom_attribute_content( $amp_story, 'poster-portrait-src' );
+
+		$data = [
+			'title'  => $title ?: '',
+			'poster' => $poster ?: '',
+		];
+
+		set_transient( $cache_key, wp_json_encode( $data ), $cache_ttl );
+
+		return rest_ensure_response( $data );
```

This should maybe return an error if an actual story wasn't found from the HTML?